### PR TITLE
New inversion methods for a Matrix

### PR DESCRIPTION
**Pull request by mbaudin47**
- The new `SquareMatrix.inverse()` and `SymmetricMatrix.inverse()` methods are created.

Example:
```python
sqMatrix = ot.SquareMatrix([[1.0, 2.0, 3.0], [3.0, 2.0, 1.0], [2.0, 1.0, 3.0]])
sqInverse = sqMatrix.inverse()
```

This is a part of PR #2330.

### CODE REVIEW

File: lib/src/Base/Type/MatrixImplementation.cxx [hunk 1]
```diff
@@ -895,6 +896,40 @@ MatrixImplementation MatrixImplementation::solveLinearSystemSquare(const MatrixI
...
 
+/* Square matrix inverse */
+MatrixImplementation MatrixImplementation::inverseSquareInPlace()
+{
+  if (nbColumns_ != nbRows_ ) throw InvalidDimensionException(HERE) << "The matrix has " << nbRows_ << " and " << nbColumns_ << ", expected a square matrix.";
+  MatrixImplementation identity(nbRows_, nbColumns_);
+  for(UnsignedInteger i = 0; i < nbRows_; ++i)
+    identity(i, i) = 1.0;
+  const MatrixImplementation inverseMatrix(solveLinearSystemSquareInPlace(identity));
+  return inverseMatrix;
```
Comment (mbaudin47) [inline lib/src/Base/Type/MatrixImplementation.cxx:907]: Comment from @regislebrun : "The method is supposed to work in-place, no?"

File: lib/src/Base/Type/MatrixImplementation.cxx [hunk 2]
```diff
@@ -895,6 +896,40 @@ MatrixImplementation MatrixImplementation::solveLinearSystemSquare(const MatrixI
...
+}
+
+/* Symmetric matrix inverse */
+MatrixImplementation MatrixImplementation::inverseSymInPlace()
+{
+  if (nbColumns_ != nbRows_ ) throw InvalidDimensionException(HERE) << "The matrix has " << nbRows_ << " and " << nbColumns_ << ", expected a square matrix.";
+  MatrixImplementation identity(nbRows_, nbColumns_);
+  for(UnsignedInteger i = 0; i < nbRows_; ++i)
+    identity(i, i) = 1.0;
+  const MatrixImplementation inverseMatrix(solveLinearSystemSymInPlace(identity));
```
Comment (mbaudin47) [inline lib/src/Base/Type/MatrixImplementation.cxx:923]: Another comment from @regislebrun : "same here"

File: lib/src/Base/Type/MatrixImplementation.cxx [hunk 3]
```diff
@@ -32,6 +32,7 @@
 #include "openturns/Os.hxx"
 #include "openturns/Exception.hxx"
 #include "openturns/Sample.hxx"
+#include "openturns/IdentityMatrix.hxx"
```
Comment (regislebrun) [inline lib/src/Base/Type/MatrixImplementation.cxx:35]: Why do you include this file? It is not used anywhere

Comment (mbaudin47) [reply lib/src/Base/Type/MatrixImplementation.cxx:35]: This is true. Fixed at 9c332e51a

File: lib/src/Base/Type/MatrixImplementation.cxx (hunk 2 above)
Comment (mbaudin47) [reply lib/src/Base/Type/MatrixImplementation.cxx]: Fixed at 149e87072

**Comment by jschueller** [comment]
Maybe we dont want to do that: we actually rarely need to compute the inverse of a matrix but only solve the linear system

**Comment by mbaudin47** [comment]
In most cases, you are right. But WE sometimes really need the inverse, for example when we want to compute the hat matrix of a linear least squares problem. The leverages are on the diagonal.

**Comment by mbaudin47** [comment]
@jschueller , @regislebrun : Can you please review this PR?

**Comment by KieranDelamotte** [comment]
@mbaudin47 
Maybe the pseudo-inverse can help ?

**Comment by mbaudin47** [comment]
> @mbaudin47 
> Maybe the pseudo-inverse can help ?

The QR or SVD decompositions can help computing the projection matrix of a linear least squares model: i derived the equations and tested that yesterday. But this cannot be done at the matrix level. It can be implemented in the LinearModel class, which can access the decomposition. Whatever the implementation, this is not a topic of the Matrix class (this PR) or the cross validation (the target PR). The implementation can be improved afterwards once the interface is defined: there is currently no LinearModelResult.getHatMatrix() available.

File: lib/src/Base/Type/MatrixImplementation.cxx [hunk 4]
```diff
@@ -895,6 +895,30 @@ MatrixImplementation MatrixImplementation::solveLinearSystemSquare(const MatrixI
...
+/* Square matrix inverse */
+MatrixImplementation MatrixImplementation::inverseSquare() const
+{
+  if (nbColumns_ != nbRows_ ) throw InvalidDimensionException(HERE) << "The matrix has " << nbRows_ << " and " << nbColumns_ << ", expected a square matrix.";
+  MatrixImplementation identity(nbRows_, nbColumns_);
+  for(UnsignedInteger i = 0; i < nbRows_; ++i)
+    identity(i, i) = 1.0;
+  MatrixImplementation A(*this);
+  const MatrixImplementation inverseMatrix(A.solveLinearSystemSquareInPlace(identity));
+  return inverseMatrix;
```
Comment (mbaudin47) [inline lib/src/Base/Type/MatrixImplementation.cxx:907]: Comment from @regislebrun :

The method is supposed to work in-place, no?

Comment (mbaudin47) [reply lib/src/Base/Type/MatrixImplementation.cxx:907]: I removed the method `MatrixImplementation::inverseSquareInPlace()`.

**Comment by mbaudin47** [comment]
@jschueller , @regislebrun : Can you please review this PR?

**Comment by jschueller** [comment]
I still dont think we should add inverse methods

**Review by jschueller — APPROVED** [review]

**Review by l-brevault — APPROVED** [review]

**Review by regislebrun — APPROVED** [review]

**Pull request merged**